File: RIONet/GANet/src/main/java/org/GANet/thread_handlers/ListenerThread.java
package org.GANet.thread_handlers;

import java.io.IOException;
import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.locks.ReentrantLock;

import org.GANet.socket_handlers.ListenerSocket;
import org.GANet.packets.Packet;
import org.GANet.packets.PacketBuilder;

/**
 * A thread that listens for incoming packets from a sender and adds them to a queue
 */
public class ListenerThread extends Thread {

    private Queue<Packet> packetQueue;
    private ListenerSocket listenerSocket;

    ReentrantLock lock;

    /**
     * create a new listener thread
     * @param port the port to listen on
     * @param packetBuilder the packet builder to use to build packets
     * @throws IOException if an error occurs while creating the listener socket
     */
    public ListenerThread(int port, PacketBuilder packetBuilder) throws IOException {
        lock = new ReentrantLock();
        packetQueue = new LinkedList<Packet>();
        listenerSocket = new ListenerSocket(port, packetBuilder);
    }

    /**
     * accept a connection from a single sender
     * @throws IOException if an error occurs while accepting the connection
     */
    public void accept() throws IOException {
        listenerSocket.accept();
    }

    public void run() {
        while (true) {
            try {
                Packet packet = listenerSocket.getPacket();
                lock.lock();

                try {
                    packetQueue.add(packet);
                } finally {
                    lock.unlock();
                }
            } catch (IOException e) {
                System.out.println("An error accured while recieving data from sender: " + e);
            }
        }
    }

    /**
     * get the next packet from the recieved packet queue
     *
     * @return a Packet from the threads packet queue, returns null if
     *         the queue is empty
     */
    public Packet getPacket() {
        lock.lock();
        try {
            return packetQueue.poll();
        } finally {
            lock.unlock();
        }
    }
}

File: PyNet/ga2230Net/thread_handlers/ListenerThread.py
from threading import Thread
from queue import Queue
import socket

from ..socket_handlers.ListenerSocket import ListenerSocket
from ..packets.Packet import Packet
from ..packets.PacketBuilder import PacketBuilder


class ListenerThread(Thread):
    """A thread handler for a listener that will continuously listen
    on a specified port and insert all data recieved into a queue
    """
    def __init__(self, port: int, packet_builder: PacketBuilder) -> None:
        """
        :param port: the port to listen on
        :type port: int
        :param packet_builder: the packet builder to use
        :type packet_builder: PacketBuilder
        """
        super().__init__()
        self.listener_socket: ListenerSocket = ListenerSocket(
            port, packet_builder
        )
        self.data_queue: "Queue[Packet]" = Queue()

        self.running = True

    def accept(self) -> None:
        """Accepts a connection from a single sender
        """
        self.listener_socket.accept()

    def listen(self, listen_count: int) -> None:
        """Starts listening for sender connections
        """
        self.listener_socket.listen(listen_count)

    def run(self) -> None:
        while self.running:
            try:
                self.data_queue.put(self.listener_socket.get_data())
            except socket.error as e:
                self.running = False

    def get_packet(self) -> Packet:
        """Retrieves a packet from the packet queue

        :return: a packet
        :rtype: Packet
        """
        return self.data_queue.get()

    def is_running(self) -> bool:
        """Returns whether the thread is running

        :return: whether the thread is running
        :rtype: bool
        """
        return self.running
